hal/riscv64/_interrupts.S

```diff
@@ -118,6 +168,50 @@
 	andi a0, a0, ~SSTATUS_SIE
 	csrw sstatus, a0
 
+	/* Check FPU status */
+	srl t0, a0, 13
+	andi t0, t0, 3
+	li t1, 2       /* FPU clean */
+	bne t0, t1, 1f
```

Comment:
I think we should restore FPU context based  on the fact it has been stored for thread at hand. Consider:

1) thread A comes into the kernel with dirty FPU - ctx is saved,
2) thread B is scheduled, FPU is clean, so no ctx restore,
3) thread B comes back into the kernel with dirty FPU - ctx is saved,
4) thread A is scheduled, FPU state is clean (kernel doesn't use it), so no context restore.

This way now thread A has FPU context of thread B.

Or am I missing something?

Reply:
It's the other way: we restore FP ctx only if status is clean. But that is also currently incorrect. Code was written according to RISC-V spec https://riscv.org/wp-content/uploads/2017/05/riscv-privileged-v1.10.pdf#page=35, but now I think it must assume we have some dedicated space for storing context, as we'd restore state that wasn't saved. I'll rewrite it a little.